Blog - Using Bazel with Rust to Build and Deploy an Application (#424)
earthly/website: blog/_posts/2023-06-19-bazel-with-rust.md

---

> title: "Using Bazel with Rust to Build and Deploy an Application"

Reviewer: Hi Sameer,
Could you please change the header image? Maybe we can have Bazel and Rust logos or something like that?

Author: done Bala.

categories:
  - Tutorials
toc: true
author: Enoch Chejieh

internal-links:
 - Rust
 - Bazel
 - Deploy
 - Applications
 - Build
---

> Building and deploying software can be a complicated and time-consuming process, especially as applications grow in size and complexity. One tool that can help simplify this process is [Bazel](https://bazel.build), an open-source build tool developed by Google. [Bazel](/blog/bazel-build) is designed to make it easy to build and test large and complex codebases and is particularly well-suited for monorepos, which are codebases that contain multiple projects or components.

Reviewer: ```suggestion
Building and deploying software can be a complicated and time-consuming process, especially as applications grow in size and complexity. One tool that can help simplify this process is [Bazel](https://bazel.build), an open-source build tool developed by Google. Bazel is designed to make it easy to build and test large and complex codebases and is particularly well-suited for monorepos, which are codebases that contain multiple projects or components.
```

> One of the key features of Bazel is its ability to speed up builds and tests. Bazel's caching and dependency analysis features facilitate fast, incremental builds. This makes it possible to quickly iterate on code changes, which can be especially useful for large teams working on a codebase. Additionally, [Bazel](/blog/monorepo-with-bazel) supports multiple languages and platforms, including [Rust](https://www.rust-lang.org), and can be extended to support new languages.

Reviewer: ```suggestion
One of the key features of Bazel is its ability to speed up builds and tests. Bazel's caching and dependency analysis features facilitate fast, incremental builds. This makes it possible to quickly iterate on code changes, which can be especially useful for large teams working on a codebase. Additionally, Bazel supports multiple languages and platforms, including Rust, and can be extended to support new languages.
```

This loads the `rule_rust_dependencies` and `rust_register_toolchains` functions. Bazel uses the `rule_rust_dependencies`  function to know what dependencies your project needs to successfully run, build, or test your application.

In the same `WORKSPACE` file, add the following line of code:

~~~{ caption="WORKSPACE"}
# WORKSPACE

# Adds the necessary dependencies for the Rust rules.
rules_rust_dependencies()
~~~

The `rust_register_toolchains` function registers the Rust toolchains with the given versions and editions you will need to use within your project.

And, in the same `WORKSPACE` file, add this line of code:

~~~{ caption="WORKSPACE"}
# ./WORKSPACE

# Registers Rust toolchains with the given versions and editions.
rust_register_toolchains(
    versions=["1.66.0"],
    
    # Specifies the Rust edition to use for the registered toolchains
    edition = "2021",
)
~~~

When another developer clones the project and runs the `bazel build` command, Bazel will check for the Rust version specified in the `rust_register_toolchains`, and if the correct version of Rust isn't already installed on the local system, Bazel will download and install it before building the project.

> > **Please note:** It's important to keep the Rust version in sync across all the developers working on the project, as different versions of Rust can cause compatibility issues and build errors.

Reviewer: ```suggestion
<div class="notice--info">
It's important to keep the Rust version in sync across all the developers working on the project, as different versions of Rust can cause compatibility issues and build errors.
</div>
```

#### Using Labels to Identify and Build Specific Targets

In addition to specifying the correct version of Rust, Bazel also uses `labels` to identify and build specific targets within the project. A `label` is a unique identifier that points to a specific target, such as a binary or a library within the project.

`Labels` have the following format: `//package:target`, where `package` refers to the package or directory containing the target and `target` refers to the specific target within that package.

Bazel uses `labels` to determine which targets to build or run as well as to resolve dependencies between targets. By specifying the correct `labels`, developers can easily build and run specific parts of their projects without having to navigate through the entire codebase.

> When building a Rust project with [Bazel](/blog/bazel-build), the following command is used for building and running build targets. Try running this command in the terminal inside your current directory:

Reviewer: ```suggestion
When building a Rust project with [Bazel](/blog/bazel-build), the following command is used for building and running build targets. Try running this command in the terminal inside your current directory:
```
```suggestion
When building a Rust project with Bazel, the following command is used for building and running build targets. Try running this command in the terminal inside your current directory:
```
